File: solutions/code-carl/ch07_backtracking/T6_LC0131_Partition.java
/**
 * Practice template for:
 * - LeetCode 131. Palindrome Partitioning
 *   https://leetcode.cn/problems/palindrome-partitioning/
 */
import java.util.ArrayList;
import java.util.List;

public class T6_LC0131_Partition {
    List<String> path = new ArrayList<>();
    List<List<String>> res = new ArrayList<>();

    public List<List<String>> partition(String s) {
        backtracking(s, 0);
        return res;
    }

    public void backtracking(String input, int startIdx) {
        if (startIdx == input.length()) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = startIdx; i < input.length(); i++) {
            String sub = input.substring(startIdx, i + 1);
            if (isValid(sub)) {
                path.add(sub);
                backtracking(input, i + 1);
                path.remove(path.size() - 1);
            }
        }
    }

    public boolean isValid(String sub) {
        int l = 0, r = sub.length() - 1;
        while (l < r) {
            if (sub.charAt(l) != sub.charAt(r)) {
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
}

File: solutions/code-carl/ch07_backtracking/t6_lc0131_partition.py
"""LeetCode 131. Palindrome Partitioning
https://leetcode.cn/problems/palindrome-partitioning/
"""

from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        path: List[str] = []
        ans: List[List[str]] = []

        def backtracking(s: str, start_idx: int) -> None:
            if start_idx == len(s):
                ans.append(path[:])
                return

            for i in range(start_idx, len(s)):
                if not is_valid(s, start_idx, i):
                    continue
                path.append(s[start_idx:i + 1])
                backtracking(s, i + 1)
                path.pop()

        def is_valid(s: str, l: int, r: int) -> bool:
            if l == r:
                return True
            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True

        backtracking(s, 0)
        return ans


if __name__ == "__main__":
    solution = Solution()
    result = sorted(tuple(x) for x in solution.partition("aab"))
    expected = sorted([("a", "a", "b"), ("aa", "b")])
    assert result == expected

    solution = Solution()
    assert solution.partition("a") == [["a"]]

    print("All tests passed.")
